### ISSUE TITLE
Lights that max out at <255 brightness cause issues

### ISSUE THREAD

**Issue by geoffwatts** [geoffwatts/cbus2zigbee#10]
When a light is turned on with a cbus switch to level 255, then is read back to 254 (its max), and that value is then populated back into cbus as the current value, a subsequent switch off of the light results in cbus2zigbee ignoring the command, unless the light is first dimmed.

**Comment by geoffwatts** [comment]
2 issues:
- Ignore not getting unset, there's some kind of race condition where you can get the light into a state where it'll just ignore cbus commands
- value_max wasn't getting set because the variable wasn't referenced correctly

**Comment by geoffwatts** [comment]
Ok what I THINK is happening with the race condition:
- I have some lights that are "chatty" and report their power usage and things once every 2 or so seconds, and that comes in with a brightness report via mqtt
- If I press a cbus button to change brightness to 0, and then one of those reports comes in concurrently, ignorecbus is set and never unset

**Comment by geoffwatts** [comment]
Localbus event sender is set to "us" when a cbus message comes from Lua, whereas it's set to "cb" when it. comes from cbus

**Comment by autoSteve** [comment]
"us" = user interface / visualisation. I'd expect "sr" for set by a resident script, and yes, "cb" is C-Bus setting.

**Comment by geoffwatts** [comment]
Here's a list of event.senders: https://forum.logicmachine.net/showthread.php?tid=1562&pid=9574#pid9574

on the SHAC, "cl" seems to be another script or the user interface, "us" seems to be the resident script, and "cb" seems to be cbus.

Maybe we set grp.sender to something custom, and then ignore anything back from ourselves?  Can do same thing with mqtt with ignoring messages from our own publisher name, and chop out the ignore logic entirely?

In the branch for this issue, I have a hacky fix for this issue - @autoSteve good to get your thoughts on fixing it properly.

**Comment by geoffwatts** [comment]
This commit has the hacky fix: bfb399fce8e4a2f73a9bf72850ae8050dda31801

**Comment by autoSteve** [comment]
There is something very weird going on here. The script checks for level changed on inbound MQTT message, and only then sets ignoreCbus.

```
        if grp.getvalue(alias) ~= level then
          if logging then log('Set '..alias..' to '..level) end
          ignoreCbus[alias] = true
          SetCBusLevel(net, app, group, level, 0)
          zigbee[alias].value = level
        end
```

**Comment by geoffwatts** [comment]
Definitely 'cl' for UI/scripts - doesn't seem to matter if it's a resident script or other.

On devices, can recommend IKEA, also all generic aliexpress temperature/humidity sensors I've tried seem to be pretty good.  I've also got some zigbee remote controls next to beds for lights (talking to HA) that work pretty nicely.

The best wired-in dimmer I've found is the Vesternet ones, also sold by ozsmartthings: https://www.ozsmartthings.com.au/products/oz-smart-zigbee-dimmer-puck-toggle - except for the max level of 254 weirdness!

**Comment by autoSteve** [comment]
I've been thinking about this.

Quite possibly the script should set C-Bus to 255 and not 254 should the max for that light be 254 and it be set to maximum brightness. i.e. 254==255. This would be C-Bus intuitive from that end, unless a C-Bus level of 254 were set, which would likely interfere but then could be force set back to 255 on the bus. Basically level 254 in C-Bus would not be able to be set for that given device.

In other words, set up to 254 for the Zigbee device, but allow C-Bus to only hit 255 and not 254.

Honestly. 254? Who came up with that brilliant idea? (Philips, I'm frowning in your direction...)
